src/Proto3/Wire/Decode.hs

```diff
@@ -127,7 +127,26 @@ data ParsedField = VarintField Word64
 -- fromList [(1,[6,3]),(2,[4])]
 --
 toMap :: [(FieldNumber, v)] -> M.IntMap [v]
-toMap kvs0 = M.fromListWith (<>) . map (fmap (:[])) . map (first (fromIntegral . getFieldNumber)) $ kvs0
+toMap kvs0 = makeMap . map (first (fromIntegral . getFieldNumber)) $ kvs0
+  where
+    makeMap :: [(Int, v)] -> M.IntMap [v]
```

Comment:
I think a simpler way to do the same thing that might be as efficient would be something like:

```haskell
makeMap = IntMap.fromListWith (<>) . map adapt . NonEmpty.groupBy fst
  where
    adapt keyValues = (fst (NonEmpty.head keyValues), toList (fmap snd keyValues))
```

In fact, in some cases this might be more efficient, such as for the sequence:

```haskell
[ (1, a), (0, b), (0, c), (0, d) ]
```

… since the final three out of order elements will be grouped before insertion into the `IntMap` instead of being inserted individually.

Reply:
This almost works, but unfortunately this case specifically needs the groups to be built in reverse. I'll add a comment to that effect

Reply:
I was able to handle this without a slowdown by just treating the GT and LT cases the same, so there's no ordering assumptions. This doesn't appear to slow anything down and does help out the unordered case.